pyiron_base/job/core.py

```diff
@@ -857,18 +857,28 @@ def __del__(self):
 
     def __getitem__(self, item):
         """
-        Get/read data from the HDF5 file or access log files.
+        Get/read data from the HDF5 file, child jobs or access log files.
 
         If the job is :method:`~.decompress`ed, item can also be a file name to
         access the raw output file of that name of the job.  See available file
         with :method:`~.list_files()`.
 
+        `item` is first looked up in this jobs HDF5 file, then in the HDF5 files of any child jobs and finally it is
+        matched against any files in the job directory as described above.
+
         Args:
             item (str, slice): path to the data or key of the data object
 
         Returns:
-            dict, list, float, int: data or data object
+            dict, list, float, int, None: data or data object; if nothing is found None is returned
         """
+        try:
```

Comment:
I agree, I don't like returning None for non-existent keys, but I also don't want to refactor this, since presumably a lot of other code depends on this. So it is what it is. I can put your suggestion in though, since it is cleaner.

Reply:
No, that's not quite my concern. I'm really not confident I understand the functionality here, so let me try to explain in full detail and let me know if I'm just being an idiot.

Suppose I had a job, `'myJob'`, that had an HDF field `'foo'` *and* a child job with the name `'foo'`. Then my understanding is that our project has  something like `'myJob_hdf5/foo'` *and* `'myJob_hdf5/foo_hdf5'`

case 1) `'myJob_hdf5/foo'` has value `None`
- Orignal code:  Children get searched first and `job['foo']` returns the child job object
- Your code: We find the `None` value in `'myJob_hdf5/foo'` but explicitly ignore it with the if clause, go on to find and return the child job.
- My suggestion: `None` is found and returned.

case 2) `'myJob_hdf5/foo'` has value `'anything_but_none'`
- Original code: Children get searched first and `job['foo']` returns the child job object
- Your code: Finds and returns `'anything_but_none'`
- My suggestion: Finds and returns `'anything_but_none'`

If nodes and groups can be guaranteed not to have the same name, then this situation never comes up, and original=your=my, and I like mine best for its brevity.

Otherwise, if we're going to let this silly situation pass silently at all, then I *like* my version best -- return node values first and if it's `None` then dammit that's a valid value. But this is really a behaviour change and we shouldn't push it without wider review.

Reply:
Ok, I understand now.  Your version is nice I think, but it changes behavior a bit more, so discussing on Friday is fine.

Reply:
Next Friday 😉